Enable stricter type checks

jamesderlin opened the pull request:
Enable the `strict-casts`, `strict-inference`, and `strict-raw-types`
options for the analyzer.  Fix missing types accordingly.

Also remove an obsolete analyzer option.

johnsonmh reviewed (approved):
These look like good improvements, please re-sync and fix any merge issues. 

LGTM

jamesderlin commented:
Updated.  I should mention that I debated whether to replace `Comparable` with `Comparable<dynamic>` or with `Comparable<Object?>`.  I initially used `Comparable<dynamic>` to avoid breaking the API, but I think that `Comparable<Object?>` should be compatible, so I switched to that.  I don't feel strongly about it either way.

jamesderlin commented:
> Run dart pub get
Resolving dependencies...
The current Dart SDK version is 2.15.1.
Because basics requires SDK version >=2.17.0 <3.0.0, version solving failed.
Error: Process completed with exit code 1.

But https://github.com/google/dart-basics/blob/master/.github/workflows/dart.yml has `google/dart:latest` ... Is "latest" not the latest development (or even stable) release? (@devoncarew Do you know?)

devoncarew commented:
It looks like the CI is provisioning `2.15.1`. The setup there - using the container image - is a little old. We generally use the `setup-dart` github action now in CIs; I can set a PR.

devoncarew commented:
https://github.com/google/dart-basics/pull/48

johnsonmh commented on pubspec.yaml:
```diff
@@ -3,12 +3,12 @@
 
 name: basics
-version: 0.8.0
+version: 0.9.0
```
is `0.9.0` right here? Feels like this is an implementation-only change, it doesn't change the public API or the performance of the internals.

Maybe `0.8.1` is more appropriate?

jamesderlin replied:
I chose 0.9.0 because I bumped the Dart SDK requirement to 2.17 for `strict-casts`, and I thought that the SDK requirement could be considered breaking.  But looking at the semver spec, I guess that's not true.  I'll change it to 0.8.1

johnsonmh replied:
Ah, makes sense, SGTM
